File: native/macos/Protocol.swift
import Foundation

private let nativeProtocolVersion = 1

enum ProtocolMethod: String {
    case status
    case requestPermission
    case snapshot
    case action
}

private struct RequestEnvelope: Decodable {
    let protocolVersion: Int
    let id: String
    let method: String

    enum CodingKeys: String, CodingKey {
        case protocolVersion = "protocol"
        case id
        case method
    }
}

private struct RequestWithParams<T: Decodable>: Decodable {
    let params: T
}

private struct SuccessEnvelope<T: Encodable>: Encodable {
    let protocolVersion = nativeProtocolVersion
    let id: String
    let result: T

    enum CodingKeys: String, CodingKey {
        case protocolVersion = "protocol"
        case id
        case result
    }
}

private struct ErrorBody: Encodable {
    let code: String
    let message: String
}

private struct ErrorEnvelope: Encodable {
    let protocolVersion = nativeProtocolVersion
    let id: String
    let error: ErrorBody

    enum CodingKeys: String, CodingKey {
        case protocolVersion = "protocol"
        case id
        case error
    }
}

final class ProtocolServer {
    private let session = ComputerSession()
    private let decoder = JSONDecoder()
    private let encoder = JSONEncoder()
    private var output = FileHandle.standardOutput

    func run(input: FileHandle, output: FileHandle) {
        self.output = output
        var buffer = Data()
        while true {
            let chunk = input.availableData
            if chunk.isEmpty { break }
            buffer.append(chunk)
            while let newline = buffer.firstIndex(of: 10) {
                let line = Data(buffer[..<newline])
                buffer.removeSubrange(...newline)
                if !line.isEmpty { handle(line) }
            }
        }
    }

    private func handle(_ data: Data) {
        let envelope: RequestEnvelope
        do {
            envelope = try decoder.decode(RequestEnvelope.self, from: data)
        } catch {
            return
        }

        do {
            guard envelope.protocolVersion == nativeProtocolVersion else {
                throw NativeFailure.unsupported(
                    "Unsupported Computer protocol version: \(envelope.protocolVersion)"
                )
            }
            guard let method = ProtocolMethod(rawValue: envelope.method) else {
                throw NativeFailure.unsupported("Unknown Computer protocol method: \(envelope.method)")
            }
            switch method {
            case .status:
                try send(envelope.id, session.status())

            case .requestPermission:
                let request: PermissionRequest = try decodeParams(data)
                session.requestPermission(request.permission)
                try send(envelope.id, session.status())

            case .snapshot:
                let request: SnapshotRequest = try decodeParams(data)
                try send(envelope.id, session.snapshot(request))

            case .action:
                let request: ActionRequest = try decodeParams(data)
                try send(envelope.id, session.action(request))
            }
        } catch let failure as NativeFailure {
            sendError(envelope.id, failure)
        } catch is DecodingError {
            sendError(envelope.id, .invalid("Invalid parameters for \(envelope.method)"))
        } catch {
            sendError(
                envelope.id,
                .internalError(error.localizedDescription)
            )
        }
    }

    private func decodeParams<T: Decodable>(_ data: Data) throws -> T {
        try decoder.decode(RequestWithParams<T>.self, from: data).params
    }

    private func send<T: Encodable>(_ id: String, _ result: T) throws {
        try emit(SuccessEnvelope(id: id, result: result))
    }

    private func sendError(_ id: String, _ failure: NativeFailure) {
        try? emit(
            ErrorEnvelope(
                id: id,
                error: ErrorBody(code: failure.code, message: failure.message)
            )
        )
    }

    private func emit<T: Encodable>(_ value: T) throws {
        output.write(try encoder.encode(value))
        output.write(Data([10]))
    }
}

File: native/windows/src/protocol.rs
use std::io::{BufRead, Write};

use serde::{de::DeserializeOwned, Deserialize, Serialize};
use serde_json::Value;

use crate::models::{ActionRequest, NativeError, SnapshotRequest};
use crate::session::ComputerSession;

pub const PROTOCOL_VERSION: u32 = 1;

#[derive(Debug, Deserialize)]
struct RequestEnvelope {
    protocol: u32,
    id: String,
    method: String,
    #[serde(default)]
    params: Value,
}

#[derive(Serialize)]
struct SuccessEnvelope<T: Serialize> {
    protocol: u32,
    id: String,
    result: T,
}

#[derive(Serialize)]
struct ErrorBody<'a> {
    code: &'a str,
    message: &'a str,
}

#[derive(Serialize)]
struct ErrorEnvelope<'a> {
    protocol: u32,
    id: &'a str,
    error: ErrorBody<'a>,
}

pub fn run(session: &mut ComputerSession) -> Result<(), NativeError> {
    let stdin = std::io::stdin();
    let mut stdout = std::io::stdout().lock();
    for line in stdin.lock().lines() {
        let line = line?;
        if line.trim().is_empty() {
            continue;
        }
        handle_line(session, &line, &mut stdout)?;
    }
    Ok(())
}

fn handle_line(
    session: &mut ComputerSession,
    line: &str,
    output: &mut impl Write,
) -> Result<(), NativeError> {
    let parsed = serde_json::from_str::<RequestEnvelope>(line);
    let request = match parsed {
        Ok(value) => value,
        Err(error) => {
            let id = serde_json::from_str::<Value>(line)
                .ok()
                .and_then(|value| value.get("id")?.as_str().map(str::to_owned))
                .unwrap_or_else(|| "unknown".to_owned());
            return write_error(
                output,
                &id,
                &NativeError::invalid(format!("Invalid protocol request: {error}")),
            );
        }
    };

    if request.protocol != PROTOCOL_VERSION {
        return write_error(
            output,
            &request.id,
            &NativeError::unsupported(format!(
                "Unsupported Computer protocol version: {}",
                request.protocol
            )),
        );
    }

    let result = match request.method.as_str() {
        "status" => write_result(output, &request.id, &session.status()),
        "snapshot" => decode::<SnapshotRequest>(request.params)
            .and_then(|params| session.snapshot(params))
            .and_then(|value| write_result(output, &request.id, &value)),
        "action" => decode::<ActionRequest>(request.params)
            .and_then(|params| session.action(params))
            .and_then(|value| write_result(output, &request.id, &value)),
        "requestPermission" => Err(NativeError::unsupported(
            "Windows does not require Computer Use permission requests",
        )),
        _ => Err(NativeError::unsupported(format!(
            "Unknown Computer protocol method: {}",
            request.method
        ))),
    };

    if let Err(error) = result {
        write_error(output, &request.id, &error)?;
    }
    Ok(())
}

fn decode<T: DeserializeOwned>(value: Value) -> Result<T, NativeError> {
    serde_json::from_value(value)
        .map_err(|error| NativeError::invalid(format!("Invalid request parameters: {error}")))
}

fn write_result<T: Serialize>(
    output: &mut impl Write,
    id: &str,
    result: &T,
) -> Result<(), NativeError> {
    let value = SuccessEnvelope {
        protocol: PROTOCOL_VERSION,
        id: id.to_owned(),
        result,
    };
    serde_json::to_writer(&mut *output, &value)
        .map_err(|error| NativeError::internal(error.to_string()))?;
    output.write_all(b"\n")?;
    output.flush()?;
    Ok(())
}

fn write_error(output: &mut impl Write, id: &str, error: &NativeError) -> Result<(), NativeError> {
    let value = ErrorEnvelope {
        protocol: PROTOCOL_VERSION,
        id,
        error: ErrorBody {
            code: error.code,
            message: &error.message,
        },
    };
    serde_json::to_writer(&mut *output, &value)
        .map_err(|error| NativeError::internal(error.to_string()))?;
    output.write_all(b"\n")?;
    output.flush()?;
    Ok(())
}
